File: java/project/security-java/src/main/java/org/tby/fourdk/security/RoleBasedStrategy.java
package org.tby.fourdk.security;

import java.util.Collections;
import java.util.LinkedHashSet;
import java.util.List;
import java.util.Set;

public class RoleBasedStrategy implements AuthorizedStrategy {
    private RoleReadRepository roleRepository;

    public RoleBasedStrategy(RoleReadRepository roleRepository) {
        this.roleRepository = roleRepository;
    }

    @Override
    public Authorization isAuthorized(Permission expectedPermission, List<Role> givenRoles) {
        if (givenRoles == null || givenRoles.isEmpty()) {
            return new Authorization(false, Collections.emptyList());
        }
        var permissions = findPermissionsForGivenRoles(givenRoles);
        return new Authorization(permissions.contains(expectedPermission), permissions);
    }

    private Set<Permission> findPermissionsForGivenRoles(List<Role> informedRoles) {
        Set<Permission> permissions = new LinkedHashSet<>();
        for (Role informedRole : informedRoles) {
            var associatedPermissions = this.roleRepository.findPermissionsForGivenRoles(informedRole);
            if (associatedPermissions != null) {
                permissions.addAll(associatedPermissions);
            }
        }
        return permissions;
    }
}

File: rust/project/security-rust/src/dddk/security/authorized_strategy_impl/role_based_strategy.rs
use std::collections::HashSet;
use std::rc::Rc;
use crate::dddk::security::authorization::Authorization;
use crate::dddk::security::authorized_strategy::AuthorizedStrategy;
use crate::dddk::security::authorized_strategy_impl::role_read_repository::RoleReadRepository;
use crate::dddk::security::permission::Permission;

pub struct RoleBasedStrategy {
    role_repository: Rc<dyn RoleReadRepository>,
}

impl RoleBasedStrategy {
    pub fn new(role_repository: Rc<dyn RoleReadRepository>) -> RoleBasedStrategy {
        RoleBasedStrategy {
            role_repository
        }
    }
}

impl AuthorizedStrategy for RoleBasedStrategy {
    fn is_authorized(&self, expected_permission: Permission, given_roles: Vec<String>) -> Authorization {
        let mut user_permissions = HashSet::new();
        given_roles.iter().for_each(|role| {
            if let Some(role) = self.role_repository.find_role_by_name(role.clone()) {
                role.get_associated_permissions().iter().for_each(
                    |permission| { user_permissions.insert(permission.clone()); }
                );
            }
        });
        if user_permissions.contains(&expected_permission) {
            return Authorization::new(true, Vec::from_iter(user_permissions));
        }
        Authorization::new(false, Vec::new())
    }
}